shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -63,13 +58,277 @@ final class TsTypegenCodeBuilder {
+      }
+
+    // add methods
+    methods.foreach { case (name, methodType) =>
+      traitDeclaration += addMethodDeclaration(traitInfo.params.toSet, name, methodType)
+    }
+
+    typegenCode += traitDeclaration + SourceCode.closeCurlyBrace
+  }
+
+  def addTypeGenClassDef(classInfo: ClassSymbol, methods: List[(String, Type)]): Unit = {
+    val className = classInfo.lexicalName
+    val classBody = classInfo.body
+    val baseClass = typeScope.resolveBaseClass(classBody)
+    val typeParams = classInfo.params.iterator.map(SourceCode(_)).toList
+
+    // create mapping for class body fields and types
+    // class body includes fields from all implemented traits
+    val classFieldsAndType = getClassFieldAndTypes(classInfo)
+    // fields can be re-declared with different types at different
+    // levels of inheritance. The final type is a union of all these types
+      .groupMap(_._1)(_._2)
+
+    // extend with base class if it exists
+    var classDeclaration = SourceCode(s"export declare class $className") ++ SourceCode.paramList(typeParams)
+    baseClass.map(baseClass => {
+      val baseClassName = baseClass.lexicalName
+      val baseClassTypeParams = baseClass.params.iterator.map(SourceCode(_)).toList
+      classDeclaration ++= SourceCode(s" extends ${baseClass.lexicalName}") ++ SourceCode.paramList(baseClassTypeParams)
+    })
+    classDeclaration ++= SourceCode.space ++ SourceCode.openCurlyBrace
+
+    // add body fields
+    classFieldsAndType
+      .map {case (fieldVar, fieldTypes) => 
+        // field types will always have 1 or more elements
+        val fieldTypesCode = if (fieldTypes.length === 1) {
+          toTsType(fieldTypes(0))(TypegenContext(fieldTypes(0)), Some(true))
+        } else {
+          // multiple types are intersected hence typgen is done
+          // using intersection precedence however only unique types
+          // are considered for intersection
+          val fieldTypesCode = fieldTypes.toSet[Type].map(fieldType => toTsType(fieldType, false, 1)(TypegenContext(fieldType), Some(true))).toList
+          SourceCode.sepBy(fieldTypesCode, SourceCode.ampersand)
+        }
+        (SourceCode(fieldVar.name), fieldTypesCode)
+      }
+      .foreach({ case (fieldVar, fieldType) => {
+        classDeclaration += SourceCode("    ") ++ fieldVar ++ SourceCode.colon ++ fieldType
+      }})
+
+    // constructor needs all fields including super classes
+    val allFieldsAndTypes = (classFieldsAndType ++
+      baseClass.map(getClassFieldAndTypes(_, true)).getOrElse(List.empty).groupMap(_._1)(_._2))
+      .map {case (fieldVar, fieldTypes) =>
+        val fieldTypesCode = fieldTypes.distinct match {
+          case Seq(fieldType) =>
+            toTsType(fieldType)(TypegenContext(fieldType), Some(true))
+          case uniqueFieldTypes =>
+            // multiple types are intersected hence typegen is done
+            // using intersection precedence however only unique types
```

Comment:
I don't understand what "using intersection precedence" means. Also, there seems to be some missing punctuation after it.

Reply:
```scala
  private def typePrecedence(mlType: Type): Int = mlType match {
    case _: Inter | _: Union => 1
    case _ => 0
  }
```

We're using operator precedence to parenthesize some parts differently, specifically functions for e.g. `number | ((arg: number) => number)`.

But I'll change the comments to make this explicit.

```
            // multiple types are intersected hence typegen is done
            // using precedence of the intersection operator however
            // only distinct types are considered for intersection
```

Reply:
Okay. But I'd write "using **the** precedence of the intersection operator".

Also not sure why "However" is used here.

By the way, this seems wrong:

> ```scala
>   private def typePrecedence(mlType: Type): Int = mlType match {
>     case _: Inter | _: Union => 1
>     case _ => 0
>   }
> ```

Intersection have precedence over unions.

Reply:
Ok I've corrected this.